dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java

```diff
@@ -702,13 +701,28 @@ final ByteResourceInformation findClassBytes(String className, String resourceNa
      * Find a class on this loader's class path or delegate to the parent class
      * loader.
      */
+    @FFDCIgnore(ClassNotFoundException.class)
     protected Class<?> findOrDelegateLoadClass(String name) throws ClassNotFoundException {
-        // The resolve parameter is a legacy parameter that is effectively
-        // never used as of JDK 1.1 (see footnote 1 of section 5.3.2 of the 2nd
-        // edition of the JVM specification).  The only caller of
-        // loadClass(String, boolean) is java.lang.ClassLoader.loadClass(String),
-        // and that method always passes false, so we ignore the parameter.
-        return super.loadClass(name, false);
+        Class<?> result = findLoadedClass(name);
+        if (result != null) {
+            return result;
+        }
+        if (parent instanceof NoClassNotFoundLoader) {
+            result = ((NoClassNotFoundLoader) parent).loadClassNoException(name);
+        } else {
+            try {
+                if (parent == null) {
```

Comment:
Should this just be the first check in the method and skip everything else if the parent is null?  Otherwise you end up calling findLoadedClass twice.

Reply:
We must check findLoadedClass before delegating to parent, otherwise we will continually delegate to parent for already defined classes.

findLoadedClass should be super fast, it should only be called twice if we have not already loaded class so I doubt it is measurable hit.  If we find it does then I suggest we have another variant of findClass that avoids getting the class loading lock and check for findLoadedClass because both are already done in this context.

Reply:
Calling super will end up calling findLoadedClass first already, just like it did before you made this change.

Reply:
If you want me to clean this up in a follow on PR that is fine, but your analysis it incorrect for sure.

Reply:
I see what you mean now.  I misunderstood you to suggest we check instance of NoClassNotFoundLoader first.  But yes we can check parent == null first here.

Reply:
I fixed this up in a follow on commit.